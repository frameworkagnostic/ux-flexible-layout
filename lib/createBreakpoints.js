'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keys = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.createBreakpoints = createBreakpoints;

var _index = require('./constants/index');

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } // Sorted ASC by size. That's important.
// It can't be configured as it's used statically for propTypes.

var keys = exports.keys = _index.BREAKPOINT_KEYS;

// Keep in mind that @media is inclusive by the CSS specification.
function createBreakpoints() {
  var breakpoints = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var _breakpoints$values = breakpoints.values,
      values = _breakpoints$values === undefined ? _index.BREAKPOINTS : _breakpoints$values,
      _breakpoints$unit = breakpoints.unit,
      unit = _breakpoints$unit === undefined ? _index.UNIT : _breakpoints$unit,
      _breakpoints$step = breakpoints.step,
      step = _breakpoints$step === undefined ? _index.STEP : _breakpoints$step,
      other = _objectWithoutProperties(breakpoints, ['values', 'unit', 'step']);

  function up(key) {
    var value = typeof values[key] === 'number' ? values[key] : key;
    return '@media (min-width:' + value + unit + ')';
  }

  function down(key) {
    var endIndex = keys.indexOf(key) + 1;
    var upperbound = values[keys[endIndex]];

    if (endIndex === keys.length) {
      // xl down applies to all sizes
      return up('xs');
    }

    var value = typeof upperbound === 'number' && endIndex > 0 ? upperbound : key;
    return '@media (max-width:' + (value - step) / 100 + unit + ')';
  }

  function between(start, end) {
    var endIndex = keys.indexOf(end) + 1;

    if (endIndex === keys.length) {
      return up(start);
    }

    return '@media (min-width:' + values[start] + unit + ') and ' + ('(max-width:' + (values[keys[endIndex]] - step) / 100 + unit + ')');
  }

  function only(key) {
    return between(key, key);
  }

  function width(key) {
    return values[key];
  }

  return _extends({
    keys: keys,
    values: values,
    up: up,
    down: down,
    between: between,
    only: only,
    width: width
  }, other);
}
//# sourceMappingURL=createBreakpoints.js.map