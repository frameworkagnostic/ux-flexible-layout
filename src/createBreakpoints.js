// Sorted ASC by size. That's important.
// It can't be configured as it's used statically for propTypes.

import {
  BREAKPOINT_KEYS,
  BREAKPOINTS,
  UNIT,
  STEP
} from './constants/index';

export const keys = BREAKPOINT_KEYS;

// Keep in mind that @media is inclusive by the CSS specification.
export function createBreakpoints(breakpoints = {}) {
  const {
    // The breakpoint **start** at this value.
    // For instance with the first breakpoint xs: [xs, sm[.
    values = BREAKPOINTS,
    unit = UNIT,
    step = STEP,
    ...other
  } = breakpoints;

  function up(key) {
    const value = typeof values[key] === 'number' ? values[key] : key;
    return `@media (min-width:${value}${unit})`;
  }

  function down(key) {
    const endIndex = keys.indexOf(key) + 1;
    const upperbound = values[keys[endIndex]];

    if (endIndex === keys.length) {
      // xl down applies to all sizes
      return up('xs');
    }

    const value = typeof upperbound === 'number' && endIndex > 0 ? upperbound : key;
    return `@media (max-width:${(value - step) / 100}${unit})`;
  }

  function between(start, end) {
    const endIndex = keys.indexOf(end) + 1;

    if (endIndex === keys.length) {
      return up(start);
    }

    return (
      `@media (min-width:${values[start]}${unit}) and ` +
      `(max-width:${(values[keys[endIndex]] - step) / 100}${unit})`
    );
  }

  function only(key) {
    return between(key, key);
  }

  function width(key) {
    return values[key];
  }

  return {
    keys,
    values,
    up,
    down,
    between,
    only,
    width,
    ...other,
  };
}
