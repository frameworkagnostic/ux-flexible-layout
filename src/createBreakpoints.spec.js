
import { keys , createBreakpoints } from './createBreakpoints';

describe('breakpoints', () => {
  test('isFunction', () => expect(typeof createBreakpoints).toBe('function'));

  describe('createBreakpoints instantiate', () => {
    const breakpoints = createBreakpoints();
    test('isObject', () => expect(breakpoints).toEqual(expect.any(Object)));
    test('snapshot', () => expect(keys).toMatchSnapshot());
  });

  describe('breakpointKeys', () => {
    test('isArray', () => expect(keys).toEqual(expect.any(Array)));
    test('snapshot', () => expect(keys).toMatchSnapshot());
  });
});
