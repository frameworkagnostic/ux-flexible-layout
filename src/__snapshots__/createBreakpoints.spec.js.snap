// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breakpoints breakpointKeys snapshot 1`] = `
Array [
  "xs",
  "sm",
  "md",
  "lg",
  "xl",
]
`;

exports[`breakpoints createBreakpoints instantiate snapshot 1`] = `
Array [
  "xs",
  "sm",
  "md",
  "lg",
  "xl",
]
`;
